src/main/java/org/opensearch/ad/settings/AbstractSetting.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.ad.settings;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+
+/**
+ * A container serving dynamic numeric setting.  The caller does not have to call
+ * ClusterSettings.addSettingsUpdateConsumer and can access the most-up-to-date
+ * value using the singleton instance. This is convenient for a setting that's
+ * accessed by various places or it is not possible to install ClusterSettings.addSettingsUpdateConsumer
+ * as the enclosing instances are not singleton (i.e. deleted after use).
+ *
+ */
+public abstract class AbstractSetting {
+    private static Logger logger = LogManager.getLogger(AbstractSetting.class);
+
+    private ClusterService clusterService;
+    /** Latest setting value for each registered key. Thread-safe is required. */
+    private final Map<String, Object> latestSettings = new ConcurrentHashMap<>();
+
+    private final Map<String, Setting<?>> settings;
+
+    protected AbstractSetting(Map<String, Setting<?>> settings) {
+        this.settings = settings;
+    }
+
+    private void setSettingsUpdateConsumers() {
+        for (Setting<?> setting : settings.values()) {
+            clusterService.getClusterSettings().addSettingsUpdateConsumer(setting, newVal -> {
+                logger.info("[AD] The value of setting [{}] changed to [{}]", setting.getKey(), newVal);
```

Comment:
here any children classes of AbstractSetting will always use the logger defined with AbstractSetting 
private static Logger logger = LogManager.getLogger(AbstractSetting.class);

Following in an example on how to use different logger for the children classes:

```
public class TestA {
    protected String log = "logA";

    private void internalMethod() {
        System.out.println(log);
    }

    public void externalMethod() {
        internalMethod();
    }
}

public class TestB extends TestA {
    public TestB() {
        log = "logB";
    }
}

public class Test {
    public static void main(String[] args) {
        TestA test = new TestB();
        test.externalMethod();
    }
}
```

Reply:
the logger is a static field and cannot be set in constructor.  So I need to put the reassignment in a static block.

```
class TestA {
    protected static String log = "logA";

    private void internalMethod() {
        System.out.println(log);
    }

    public void externalMethod() {
        internalMethod();
    }
}
 class TestB extends TestA {
     {
         log = "logB";
     }
    public TestB() {
    }
}

public class Main {
    public static void main(String[] args) {
        TestA test = new TestB();
        test.externalMethod();
    }
}
```

I can use the setting name to differentiate which subclass it is.  Will use your technique in the future when needed :)

Reply:
Cool!